hibernate/cache-v53/src/main/java/org/infinispan/hibernate/cache/v53/impl/NonTxInvalidationCacheAccessDelegate.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later.
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.infinispan.hibernate.cache.v53.impl;
+
+import java.util.concurrent.CompletableFuture;
+
+import org.hibernate.engine.spi.SharedSessionContractImplementor;
+import org.infinispan.commands.write.AbstractDataWriteCommand;
+import org.infinispan.commands.write.RemoveCommand;
+import org.infinispan.context.InvocationContext;
+import org.infinispan.context.impl.FlagBitSets;
+import org.infinispan.hibernate.cache.commons.InfinispanDataRegion;
+import org.infinispan.hibernate.cache.commons.access.PutFromLoadValidator;
+
+/**
+ * Delegate for non-transactional invalidation caches
+ *
+ * @author Radim Vansa &lt;rvansa@redhat.com&gt;
+ */
+class NonTxInvalidationCacheAccessDelegate extends org.infinispan.hibernate.cache.commons.access.NonTxInvalidationCacheAccessDelegate {
+   public NonTxInvalidationCacheAccessDelegate(InfinispanDataRegion region, PutFromLoadValidator validator) {
+      super(region, validator);
+   }
+
+   @Override
+   protected void registerRemoteInvalidation(AbstractDataWriteCommand command, Object session) {
+      Sync sync = (Sync)((SharedSessionContractImplementor) session).getCacheTransactionSynchronization();
+      if (trace) {
+         log.tracef("Registering synchronization on transaction in %s, cache %s: %s", command.getKeyLockOwner(), cache.getName(), command.getKey());
+      }
+      sync.registerAfterCommit(new InvalidationInvocation(nonTxPutFromLoadInterceptor, command.getKey(), command.getKeyLockOwner()));
+   }
+
+   @Override
+   protected void invoke(Object session, InvocationContext ctx, RemoveCommand command) {
+      CompletableFuture<Object> future = invoker.invokeAsync(ctx, command);
+      // LockingInterceptor removes this flag when the command is applied and we only wait for RPC
+      if (!command.hasAnyFlag(FlagBitSets.FORCE_WRITE_LOCK)) {
```

Comment:
I'm not sure if this works. The flag is removed when the command is completed and, if I understand completely, it only happens after the rpc is completed.

Reply:
similar comment for `TombstoneAccessDelegate`

Reply:
The flag is removed when the command is applied (to DC) and the key is unlocked. At that point the command is not completed yet, the completion is waiting for the RPC.

This is an optimization based on the assumption that the command does not get blocked by locking and gets applied to DC in this thread, so when we get the future, it's already written to DC. In this case the `FWL` is unset and we don't have to block, as here we don't need to wait until the RPC completes: all we care about is that it is locally applied. We'll have to eventually wait until the RPC completes (int the synchronization), but we can proceed with other commands.

If we happen to be unlucky and have to wait for the lock, we get the future before the command is applied. We can't wait specifically for the local application, so we wait for the RPC too (and this gets completed only after local application).

Reply:
I didn't really follow the discussion, but this sounds like something that would be useful in a comment :)

Reply:
what cache mode is used? INVALIDATION?
I think the optimized path isn't hit in repl or dist mode (except if you are the primary owner)

in any case, I would prefer a better future proof solution. If we introduce the flag removal in another interceptor, you can start observing some weird behaviour. 

but it is up to you :+1: (I guess you could convert the regular operation to function operation?)

Reply:
The lock is acquired on all owners; so in this case we acquire the lock on originator and then the command is sent to all other owners, the order in which we apply invalidations doesn't matter so we don't need to go through primary owner. Locking interceptor expect the CF from NonTxII and waits for that after releasing the lock.

I agree that this is not future proof... I could consider using functional remove as the solution there is still hackish but less prone to be broken. Any 'proper' (design-wise) solution would increase size of the command for the general execution (-1) or introduce new flags that will reflect a 'status' of the command.

Reply:
@rvansa I think it'd be good to have a design or how-it-works document in the 2LC module that explains all this in detail. The 2LC module works quite differently to the rest of the codebase, so it's common to see these kind of questions. In essence, I think we should cleanup/update explanations [here](https://github.com/infinispan/infinispan-designs/blob/master/Infinispan-Hibernate-Second-Level-Cache-improvements.md) and add it to the module itself as part of the readme or similar doc. I don't think this should be in the user guide but it's useful for contributors. Now that the 2LC modules are part of the Infinispan codebase, there are more eyes on this :)